CDB: file column type QOL

Yanrishatum opened the pull request:
* Added drag&drop support for file column.
* Added a layout toggle to inline image files directly into table instead of having them hover-on. Default off, saved with layout globally.
* Fixed preview positioning when table scroll position is not 0
* Restyled preview window a bit to make it more fitting with dark theme of HIDE.

Things to consider:
* I'm not sure of the toggle storage location in global config.
* Also not sure if toggle menu item should be in Database or Layout.

Yanrishatum commented:
Found and fixed a bug with preview positioning when working with tables that span more than one page and you have to scroll.
Also changed preview colors from bright to dark, as they were sticking out of general Hide dark color scheme.

ncannasse commented:
CDB display options should instead be handled through hide props.json system.
You can use the following keys:

```
{
   "cdb" : {  /* all sheets options */ },
   "cdb.mysheet" : { /* specific options for mysheet */ }
}
```

Yanrishatum commented:
Which config it should use? (global/project/user/current) Also I would like to avoid the need for user to manually edit props.json when using that feature. Also I'm not fond of keeping it in project props.json as it may be shared between multiple users, and some may not want it enabled.

ncannasse commented:
For "reading" the config, it should always be in current, this way users can enable it at any level, either per project or in their own props.json

For documentation purposes you can set a default value in hide `bin/defaultProps.json`

We currently don't have a way to do configuration other than manually editing the json. That would require something similar to VSCode that adds metadata for each parameter and generate an user interface from it that edits the JSON in background. Until we have this I prefer to keep configuration the current way and not UI specific to a given configuration.

Yanrishatum commented:
Changed config location to use props json file, added docs to defaultProps.json, removed toggle from menus. I decided against allowing listing multiple columns in config (`"cdb.mysheet.mycol1.mycol2.mycol3"`), since it would introduce unnecessary complexity in finding that entry. So it's only global, for entire sheet and for one column.

ncannasse commented:
Almost there ;)

For sub sheets (for instance an File in a Property) the sub sheet name will be sheet@column, so you need to sheet.name.split("@").join(".") to get the sheet correct path for config lookup (don't use sheet.path as this also includes the list line number).

Yanrishatum commented:
Since subsheet depth is unknown, opted to using `while` in order to find the config value.

kLabz commented:
I just tried this branch (with `cdb.inlineImageFiles = true`) and I couldn't edit my [image] file columns via double click any more for some reason.

Also, I was expecting drag&drop to work from Resources panel, could it work too?

Yanrishatum commented:
Fixed double-click.

And on drag&drop from Resources - I'm not confident with Hide enough to get it working. Mainly this is because CDB editor is not a View, hence some part of it should become one to receive drag&drop from Resources panel.

kLabz commented on hide/comp/cdb/Cell.hx:
```diff
@@ -30,6 +30,20 @@ class Cell extends Component {
+					refresh();
+				}
+			});
+			if ( canEdit() )
```
Shouldn't the drop listener be in `if ( canEdit() )` too btw?

Yanrishatum replied:
I'm actually unsure. @ncannasse I'm not sure how `canEdit()` determines editability. Should TFile check for it, or be editable always? And if it have to be enforced, shouldn't TScript also check for `canEdit()`?

ncannasse replied:
Yes canEdit() should be taken into account. It's a new feature to allow partial edition so is not yet done everywhere.

ncannasse commented:
Please fix the canEdit() so I can merge ;)
